File: lib/rag/parse.ts
import mammoth from "mammoth";
import type { PageText } from "./chunk";

/** Parse a supported file into page-segmented text. */
export async function parseFile(
  buffer: Buffer,
  filename: string,
  mime: string,
): Promise<PageText[]> {
  const ext = filename.toLowerCase().split(".").pop() ?? "";

  if (ext === "pdf" || mime === "application/pdf") {
    return parsePdf(buffer);
  }
  if (
    ext === "docx" ||
    mime === "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
  ) {
    const { value } = await mammoth.extractRawText({ buffer });
    return [{ page: 1, text: value }];
  }
  if (["txt", "md", "markdown", "csv", "json"].includes(ext) || mime.startsWith("text/")) {
    return [{ page: 1, text: buffer.toString("utf8") }];
  }

  throw new Error(`Format tidak didukung: .${ext}. Gunakan PDF, DOCX, TXT, atau MD.`);
}

async function parsePdf(buffer: Buffer): Promise<PageText[]> {
  // pdf-parse v2 ships ESM; import dynamically so it stays out of the bundle.
  const { PDFParse } = await import("pdf-parse");
  const parser = new PDFParse({ data: new Uint8Array(buffer) });
  try {
    const result = await parser.getText();
    return result.pages
      .map((p) => ({ page: p.num, text: p.text }))
      .filter((p) => p.text.trim().length > 0);
  } finally {
    await parser.destroy();
  }
}

export const ACCEPTED_EXTENSIONS = [".pdf", ".docx", ".txt", ".md", ".csv", ".json"];

File: backend/app/rag/parse.py
"""Parse uploaded files into page-segmented text."""

import io
import re

DOCX_MIME = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
TEXT_EXTS = {"txt", "md", "markdown", "csv", "json"}


def ext_of(filename: str) -> str:
    ext = filename.lower().rsplit(".", 1)[-1] if "." in filename else "bin"
    return re.sub(r"[^a-z0-9]", "", ext) or "bin"


def parse_file(buffer: bytes, filename: str, mime: str) -> list[dict]:
    ext = ext_of(filename)

    if ext == "pdf" or mime == "application/pdf":
        return _parse_pdf(buffer)
    if ext == "docx" or mime == DOCX_MIME:
        return _parse_docx(buffer)
    if ext in TEXT_EXTS or mime.startswith("text/"):
        return [{"page": 1, "text": buffer.decode("utf-8", errors="replace")}]

    raise ValueError(f"Unsupported format: .{ext}. Use PDF, DOCX, TXT, or MD.")


def _parse_pdf(buffer: bytes) -> list[dict]:
    from pypdf import PdfReader

    reader = PdfReader(io.BytesIO(buffer))
    pages: list[dict] = []
    for i, page in enumerate(reader.pages):
        text = page.extract_text() or ""
        if text.strip():
            pages.append({"page": i + 1, "text": text})
    return pages


def _parse_docx(buffer: bytes) -> list[dict]:
    import docx

    document = docx.Document(io.BytesIO(buffer))
    text = "\n".join(p.text for p in document.paragraphs)
    return [{"page": 1, "text": text}]
